doc/source/train/user_guide.rst

```diff
@@ -497,6 +497,35 @@ A simple example for creating a callback that will print out results:
     trainer.shutdown()
 
 
+Results Preprocessors
```

Comment:
Discussed offline, but I think the benefit of preprocessors here is that it allows you to easily customize existing callbacks, rather than for use when developing custom callbacks.

Instead of the example below, I would change this to 

```python
callback = PrintingCallback()
callback.results_preprocessor = IndexedResultsPreprocessor(0)
```

And I would also move this subsection to the "Built-in callbacks" (and rename to "Advanced: Customizing Built-in Callbacks") instead of the "Custom Callbacks" Section.

Then in the "Custom Callbacks" section I would add a note linking to this section if you want to customize existing callbacks.

Reply:
Actually looking at the example above, in my opinion, it would be cleaner if callbacks would accept an optional list of result preprocessors in the `__init__`, rather than having to set an attribute.

Reply:
If we want to make this a developer API, then I think we can actually not include this information in the user guide, right? (we can still leave it here, but comment it out).